### PR TITLE
vhci - multiple interfaces, "usbip install" feature

### PR DESCRIPTION
**Pull request by koudis**
**PR which repair some bug and improve support for composite interfaces. Support for `usbip install` feature**

I had problems with usb devices which has many (>10) interfaces.

There were several problems in HUB implementation:

- IoCancelIrp - bug in Dpc release lock 
- ABORT_PIPE request was not implemented correctly.
- RESET_PIPE_AND_CLEAR_STALL was not implemented correctly against USBIP server hosted on linux machine. I guess that RESET_PIPE_AND_CLEAR_STALL works fine for USBIP server hosted on Windows machine.
- CONTROL_TRANSFER_EX was NOT implement

Problems in usbip.exe userspace program:

- I updated `usbip_forward` routine --> no deadlock, segfaults or BSOD is now possible (I tested it for vhci not for stub driver)
- Change process priority to HIGH for `attach` command

I also merge StoyanDimitrov changes to my brach (due to small update in README.md).

After the PR test and approve I will merge all commits into one.

### CODE REVIEW

**Comment by koudis** [comment]
I see that there are some changes which needs to be clean up.

File: driver/vhci/vhci_read.c [hunk 1]
```diff
@@ -79,30 +82,44 @@ store_urb_reset_dev(PIRP irp, struct urb_req *urbr)
 static NTSTATUS
 store_urb_reset_pipe(PIRP irp, PURB urb, struct urb_req *urbr)
 {
+	/* 1. We clear STALL/HALT feature on endpoint specified by pipe
```
Comment (koudis) [inline driver/vhci/vhci_read.c]: This function can be problematic - It works with USBIP server under linux, but USBIP server under windows handle USB_REQUEST_RESET_PIPE explicitly.

**Comment by koudis** [comment]
@cezanne It seem that changes works well.
I've tested it for 3 weeks and everithing works fine.

Can you review my PR please?

(There is problem with audio and other devices which rely on proper timing but it's not due to this PR)

**Comment by cezanne** [comment]
@koudis : sorry for my long silence. I'll investigate your great work and try to merge.

**Comment by cezanne** [comment]
@koudis : [koudis_install branch](/cezanne/usbip-win/tree/koudis_install) is created from your commits(02cef87 .. 45d3f1d). I slightly modified your commit to remove multiple blank lines and usbip_install_usage typo error. I know that the typo error was  resolved by your recent commit. 

Please check a merged commit ( 55c859159856040c4f6736b4fe0e1f7372af50a4 )
 
Now, I would like to merge remaining your work related to a multiple interfaces issue.

**Comment by koudis** [comment]
Just warning: commits between f974e8eeeaee939ac6a59c1c539ea5f37a13c90b and the HEAD of my branch is not many useful :).

There are also some changes for Install feature is not between 02cef87 .. 45d3f1d
- 16b75be887b44a8226f496da4ca8d2eaad177972
- 57384bdb8625d14277e4cf2dc861cf56d8596abb
- 3e36af704a70be116871ed35bc89a46e1479fb85
- 29776972353797918f4d83d6a20a5225a9355cbb

But it seems that you already merged these commits into koudis_instal :).

I will test it tomorrow. (2. 2. or 3. 2.)

**Comment by cezanne** [comment]
> Just warning: commits between [f974e8e](https://github.com/cezanne/usbip-win/commit/f974e8eeeaee939ac6a59c1c539ea5f37a13c90b) and the HEAD of my branch is not many useful :).
> 
> Just for sure - there are also some changes for Install feature which is not between [02cef87](https://github.com/cezanne/usbip-win/commit/02cef87b030adc52a7fd850bf7915e89e09f5622) .. [45d3f1d](https://github.com/cezanne/usbip-win/commit/45d3f1dbbbefadbe4b9bb921675a09d5e10bfd22)
> 
> * [16b75be](https://github.com/cezanne/usbip-win/commit/16b75be887b44a8226f496da4ca8d2eaad177972)
> * [57384bd](https://github.com/cezanne/usbip-win/commit/57384bdb8625d14277e4cf2dc861cf56d8596abb)
> * [3e36af7](https://github.com/cezanne/usbip-win/commit/3e36af704a70be116871ed35bc89a46e1479fb85)
> * [2977697](https://github.com/cezanne/usbip-win/commit/29776972353797918f4d83d6a20a5225a9355cbb)
> 
> But it seems that you already merged these commits into koudis_instal :).
> 
> I will test it tomorrow. (2. 2. or 3. 2.)

**Comment by cezanne** [comment]
@koudis : I've just closed by mistake and reopened.
I have tried to upload 2 commits from your works (one for usbip.exe and another for vhci driver).
Indeed, your commits are too~ long. 😄 

> Just warning: commits between [f974e8e](https://github.com/cezanne/usbip-win/commit/f974e8eeeaee939ac6a59c1c539ea5f37a13c90b) and the HEAD of my branch is not many useful :).
> 
> Just for sure - there are also some changes for Install feature which is not between [02cef87](https://github.com/cezanne/usbip-win/commit/02cef87b030adc52a7fd850bf7915e89e09f5622) .. [45d3f1d](https://github.com/cezanne/usbip-win/commit/45d3f1dbbbefadbe4b9bb921675a09d5e10bfd22)
> 
> * [16b75be](https://github.com/cezanne/usbip-win/commit/16b75be887b44a8226f496da4ca8d2eaad177972)
> * [57384bd](https://github.com/cezanne/usbip-win/commit/57384bdb8625d14277e4cf2dc861cf56d8596abb)
> * [3e36af7](https://github.com/cezanne/usbip-win/commit/3e36af704a70be116871ed35bc89a46e1479fb85)
> * [2977697](https://github.com/cezanne/usbip-win/commit/29776972353797918f4d83d6a20a5225a9355cbb)
> 
> But it seems that you already merged these commits into koudis_instal :).
> 
> I will test it tomorrow. (2. 2. or 3. 2.)

**Comment by koudis** [comment]
@cezuni I thought that we merge both features together. So I did not bother about "nice" commits.

If the problem with messy history occurred let me know and I will try it repair shortly :) .

File: driver/vhci/usb.ids [hunk 2]
```diff
@@ -0,0 +1,19642 @@
+#
```
Comment (cezanne) [inline driver/vhci/usb.ids]: `drivers/vhci/usb.ids` is required? We already have usb.ids at `userspace/usb.ids`.

File: driver/vhci/usbip_vhci.vcxproj [hunk 3]
```diff
@@ -77,13 +77,15 @@
...
-    <TargetVersion>Windows10</TargetVersion>
+    <TargetVersion>
+    </TargetVersion>
     <UseDebugLibraries>true</UseDebugLibraries>
     <PlatformToolset>WindowsKernelModeDriver10.0</PlatformToolset>
     <ConfigurationType>Driver</ConfigurationType>
     <DriverType>KMDF</DriverType>
     <DriverTargetPlatform>Desktop</DriverTargetPlatform>
     <SpectreMitigation>false</SpectreMitigation>
+    <_NT_TARGET_VERSION>0xA000007</_NT_TARGET_VERSION>
```
Comment (cezanne) [inline driver/vhci/usbip_vhci.vcxproj]: This property does not seem to be supported by VS 2017. Without this change, VS 2017 is also OK. 
I would drop this change if there's no special reason.

File: driver/vhci/usb.ids (hunk 2 above)
Comment (koudis) [reply driver/vhci/usb.ids]: No is not. It's a mistake. Can be removed.

File: driver/vhci/usbip_vhci.vcxproj (hunk 3 above)
Comment (koudis) [reply driver/vhci/usbip_vhci.vcxproj]: This force target to Windows 10. It's not necessary. Driver should work fine.
But I suggest to test driver well before we remove this setting. (I can do it in our server farm)

**Comment by koudis** [comment]
@cezuni i test koudis_install branch and create #98 . There were smal bugfix due to missing deps for linker.

**Comment by cezanne** [comment]
@koudis : Please check 5530943450c6030504fa2ea676152fabc56e5e59 in master branch. KeReleaseSpinLock is replaced with KeReleaseSpinLockFromDpcLevel. In 799003000389bf967543ae93656d7008e269ee38, you gave a hint. Thanks. 
I think that KeReleaseSpinLockFromDpcLevel would be more desirable than KeReleaseSpinLockForDpc. Isn't it?

**Comment by koudis** [comment]
@cezuni Yes, I've read documentation once again and the doc explicitly say
"The caller should release the spin lock with KeReleaseSpinLockFromDpcLevel as quickly as possible."

You commit looks more correct according to doc.

**Comment by cezanne** [comment]
@koudis : I created [abort_pipe branch](/cezanne/usbip-win/tree/abort_pipe) and PR #101 in order to reviewing vhci_ioctl_abort_pipe and store_urb_reset_pipe. How about moving on #101?

**Comment by koudis** [comment]
@cezuni ok :).

Are you going to merge "URB_FUNCTION_CONTROL_TRANSFER_EX" into master?

If not I suggest to test master branch agains multiple inteface devices like Logitech Unyfing device. I had problem with interface initialization without this call (but for real I am not sure if TRANFER_EX it's really necessary).

**Comment by koudis** [comment]
Maybe we can consider to check URB_FUNCTION_CONTROL_TRANSFER too due to #102

**Comment by cezanne** [comment]
> Are you going to merge "URB_FUNCTION_CONTROL_TRANSFER_EX" into master?

Yes, sure. But after #101.

> If not I suggest to test master branch agains multiple inteface devices like Logitech Unyfing device. I had problem with interface initialization without this call (but for real I am not sure if TRANFER_EX it's really necessary).

I'll keep in mind. 😄

**Pull request closed without merging**